ogre2/src/Ogre2Scene.cc

```diff
@@ -166,6 +187,164 @@ void Ogre2Scene::PreRender()
+  {
+    ignwarn << "Calling Scene::PostRender but "
+               "SetCameraPassCountPerGpuFlush is 0 (legacy mode for clients"
+               " not calling PostRender)."
+               "Read the documentation on SetCameraPassCountPerGpuFlush, "
+               "you very likely want to increase this number" << std::endl;
+  }
+  else
+  {
+    if (dataPtr->currNumCameraPasses > 0u)
+      FlushGpuCommandsAndStartNewFrame(0u, true);
+    else
+    {
+      // Every camera already calls FlushGpuCommandsAndStartNewFrame(false)
+      // right after rendering. So likely commands are already flushed.
+      //
+      // If we're here then we are only missing to perform the last step of
+      // FlushGpuCommandsAndStartNewFrame in order to start a new frame
+      EndFrame();
+    }
+  }
+}
+
+//////////////////////////////////////////////////
+void Ogre2Scene::StartForcedRender()
+{
+  if (this->LegacyAutoGpuFlush() || !this->dataPtr->frameUpdateStarted)
+  {
+    this->ogreSceneManager->updateSceneGraph();
+  }
+}
+
+//////////////////////////////////////////////////
+void Ogre2Scene::EndForcedRender()
+{
+  dataPtr->currNumCameraPasses = 0;
+  FlushGpuCommandsOnly();
+
+  if (this->LegacyAutoGpuFlush() || !this->dataPtr->frameUpdateStarted)
+  {
+    auto engine = Ogre2RenderEngine::Instance();
+    auto ogreRoot = engine->OgreRoot();
+
+    auto itor = ogreRoot->getSceneManagerIterator();
+    while (itor.hasMoreElements())
+    {
+      Ogre::SceneManager *sceneManager = itor.getNext();
+      sceneManager->clearFrameData();
+    }
+  }
+}
+
+//////////////////////////////////////////////////
+void Ogre2Scene::StartRendering()
+{
+  if (this->LegacyAutoGpuFlush())
+  {
+    auto engine = Ogre2RenderEngine::Instance();
+    engine->OgreRoot()->_fireFrameStarted();
+
+    this->ogreSceneManager->updateSceneGraph();
+  }
+  else
+  {
+    IGN_ASSERT( this->dataPtr->frameUpdateStarted == true,
+                "Started rendering without first calling Scene::PreRender. "
+                "See Scene::SetCameraPassCountPerGpuFlush for details");
+  }
+}
+
+//////////////////////////////////////////////////
+void Ogre2Scene::FlushGpuCommandsAndStartNewFrame(uint8_t _numPasses,
+                                                  bool _startNewFrame)
+{
+  dataPtr->currNumCameraPasses += _numPasses;
+
+  if (dataPtr->currNumCameraPasses >= dataPtr->cameraPassCountPerGpuFlush ||
+      _startNewFrame)
+  {
+    dataPtr->currNumCameraPasses = 0;
+    FlushGpuCommandsOnly();
+
+    // Legacy mode requires to do EndFrame here every time
+    if (dataPtr->cameraPassCountPerGpuFlush == 0u || _startNewFrame)
+      EndFrame();
+  }
+}
+
+//////////////////////////////////////////////////
+void Ogre2Scene::FlushGpuCommandsOnly()
+{
+  auto engine = Ogre2RenderEngine::Instance();
+  auto ogreRoot = engine->OgreRoot();
+  Ogre::CompositorManager2 *ogreCompMgr = ogreRoot->getCompositorManager2();
+
+  // engine->OgreRoot()->renderOneFrame();
```

Comment:
remove ?

Reply:
Decided to better document this.

That line is very useful, specially if debugging "what should happen when you let Ogre do anything" vs "what happens when we do some lower level calls ourselves"